### ISSUE TITLE
Give rules a way to coerce input data to match the rule condition

### ISSUE THREAD

**Issue by kentquirk** [honeycombio/refinery#541]
**Is your feature request related to a problem? Please describe.**

Refinery's rules engine has conditions that look like this:

```
    [[production.rule.condition]]
      field = "http.status_code"
      operator = "="
      value = 200
```

The problem is that different backends can and do send status code as "200" instead of 200. This will cause this condition to fail in refinery, but then, assuming the data is sent to Honeycomb, the Honeycomb backend would then coerce "200" to 200 (assuming the field is defined as an integer in the dataset schema). 

A team using a mixed technology stack can (and has!) run into this problem, and looking at the data in Honeycomb yields no clue, since the different datatypes have been coerced on ingestion. 

**Describe the solution you'd like**

Expand the conditions to add another element:
```
    [[production.rule.condition]]
      field = "http.status_code"
      operator = "="
      value = 200
      datatype = "int"
```

If the datatype is specified, then the value will be stored as the given datatype, and all input data will be coerced to the specified datatype before the comparison.

Like Honeycomb's backend:
* the datatypes will be limited to `int`, `float`, `string`, and `bool`.
* conversion failure results in the compare function returning false (no matter what operator is used)
* existence tests are not affected by the datatype
* bonus feature: we could add a new operator `hastype` that would compare the type of the incoming field to one of the values above.

For backwards compatibility, if datatype is not specified, no change to existing behavior will occur. 

**Describe alternatives you've considered**

Currently, if this problem is occurring, debugging is difficult, and the only solution is to add additional rules to handle the other datatypes. 

We could enable additional config that would generate debugging info about incoming datatypes, but this expands an already complex configuration in addition to the additional rule complexity.

**Additional context**

**Comment by JamieDanielson** [comment]
Does this supersede #452 ? If so we can close that one as this has much more detail.

**Comment by kentquirk** [comment]
addressed by #585
